test/distributed/org/apache/cassandra/distributed/test/streaming/BoundExceptionTest.java

```diff
@@ -0,0 +1,144 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.distributed.test.streaming;
+
+import java.io.IOException;
+import java.util.HashSet;
+
+import com.google.common.base.Throwables;
+import org.junit.Test;
+
+import org.apache.cassandra.streaming.StreamSession;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+public class BoundExceptionTest
+{
+    private static final int LIMIT = 2;
+
+    @Test
+    public void testSingleException() {
+        try {
+            throw new RuntimeException("test exception");
+        } catch (Exception e) {
+            StreamSession.boundStackTrace(e, LIMIT, new HashSet<>());
+            String expectedStackTrace = "java.lang.RuntimeException: test exception\n" +
+                                        "\tat org.apache.cassandra.distributed.test.streaming.BoundExceptionTest.testSingleException(BoundExceptionTest.java:38)\n" +
+                                        "\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n";
+            assertEquals(expectedStackTrace,Throwables.getStackTraceAsString(e));
+            assertEquals(e.getStackTrace().length, LIMIT);
+        }
+    }
+
+    @Test
+    public void testNestedException() {
+        try {
+            throw new RuntimeException(new IllegalArgumentException("the disk /foo/var is bad", new IOException("Bad disk somewhere")));
+        } catch (Exception e) {
+            StreamSession.boundStackTrace(e, LIMIT, new HashSet<>());
+            String expectedStackTrace = "java.lang.RuntimeException: java.lang.IllegalArgumentException: the disk /foo/var is bad\n" +
+                                        "\tat org.apache.cassandra.distributed.test.streaming.BoundExceptionTest.testNestedException(BoundExceptionTest.java:52)\n" +
+                                        "\tat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\n" +
+                                        "Caused by: java.lang.IllegalArgumentException: the disk /foo/var is bad\n" +
+                                        "\t... " + LIMIT + " more\n" +
+                                        "Caused by: java.io.IOException: Bad disk somewhere\n" +
+                                        "\t... " + LIMIT + " more\n";
+            assertEquals(expectedStackTrace,Throwables.getStackTraceAsString(e));
+            assertEquals(e.getStackTrace().length, LIMIT);
+        }
+    }
+
+    @Test
+    public void testExceptionCycle()
+    {
+        try {
+            throw new Exception("Test exception 1");
+        }
+        catch (Exception e1) {
+            try {
+                throw new RuntimeException("Test exception 2");
+            }
+            catch (Exception e2) {
+                e1.initCause(e2);
+                e2.initCause(e1);
+
+                causeCycle(e1);
```

Comment:
this is confusing name, you just caused a cycle... this method does not do cause one...